/*
 * @author Yanzhe Lee [lee.yanzhe@yanzhe.org]
 */
syntax = "proto3";
package com.hitnslab.dnssecurity.deeparcher.api.proto;
option java_package = "com.hitnslab.dnssecurity.deeparcher.api.proto.generated.java";
option java_outer_classname = "GraphAttributesProto";
option optimize_for = SPEED;
option cc_enable_arenas = true;

message ComponentAttr {

  /* Component id. */
  uint32 component_id = 2;

  /* Node IDs of this component in parent graph.
   * NULLABLE. */
  repeated uint32 components = 3;
}

/* General type for representing a category of graph element attribute that applies to some nodes or edges in the graph.
 * For example,
 * GraphElementsAttrMap {
 *    name: "score",
 *    target: NODE,
 *    data: { 1: 0.1, 4: 2.5  },
 *    type_url: "type.googleapis.com/google.protobuf.FloatValue"
 * }
 * means an attribute `score` of type 'google.protobuf.FloatValue' applies to node 1 and 4 with value 0.1 and 2.5.
 * By using this representation, nodes without this attribute will not occupy space. */
message GraphElementsAttrMap {

  /* Node ID to serialized attribute value mapping.
   * NOT NULL. */
  map<uint32, bytes> data = 1;

  /* Type URL of the data map values.
   * It assumes that every value of data map is a [google.protobuf.Any] object with this type_url.
   * @see [google.protobuf.Any#type_url] for more information.
   * Actually, the value of data map can be any serializable objects with same type.
   * You can omit this field if map values are not protobuf messages, and you should parse these bytes by yourself.
   * NULLABLE. */
  string type_url = 2;
}
